File: src/sdks/dotnet/Infrastructure/Aws/AwsSsmSecretProvider.cs
namespace Envilder.Infrastructure.Aws;

using Amazon.SimpleSystemsManagement;
using Amazon.SimpleSystemsManagement.Model;
using Envilder.Domain.Ports;
using System;
using System.Threading;
using System.Threading.Tasks;

/// <summary>
/// <see cref="ISecretProvider"/> backed by AWS Systems Manager Parameter Store.
/// Parameters are retrieved with decryption enabled so that <c>SecureString</c>
/// values are returned in plain text.
/// </summary>
public class AwsSsmSecretProvider : ISecretProvider
{
	private readonly IAmazonSimpleSystemsManagement _ssmClient;

	/// <summary>
	/// Initializes a new instance using the supplied SSM client.
	/// </summary>
	/// <param name="ssmClient">A configured <see cref="IAmazonSimpleSystemsManagement"/> instance.</param>
	public AwsSsmSecretProvider(IAmazonSimpleSystemsManagement ssmClient)
	{
		_ssmClient = ssmClient ?? throw new ArgumentNullException(nameof(ssmClient));
	}

	/// <inheritdoc />
	public async Task<string?> GetSecretAsync(string name, CancellationToken cancellationToken = default)
	{
		if (string.IsNullOrWhiteSpace(name))
		{
			throw new ArgumentException("Secret name cannot be null or whitespace.", nameof(name));
		}

		try
		{
			var response = await _ssmClient.GetParameterAsync(new() { Name = name, WithDecryption = true }, cancellationToken);
			return response.Parameter.Value;
		}
		catch (ParameterNotFoundException)
		{
			return null;
		}
	}

	/// <inheritdoc />
	public string? GetSecret(string name)
	{
		if (string.IsNullOrWhiteSpace(name))
		{
			throw new ArgumentException("Secret name cannot be null or whitespace.", nameof(name));
		}

		try
		{
			using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(60));
#pragma warning disable VSTHRD002 // AWS SDK v4 has no synchronous GetParameter API
			var response = _ssmClient.GetParameterAsync(new() { Name = name, WithDecryption = true }, cts.Token).GetAwaiter().GetResult();
#pragma warning restore VSTHRD002
			return response.Parameter.Value;
		}
		catch (ParameterNotFoundException)
		{
			return null;
		}
	}
}

File: src/sdks/nodejs/src/infrastructure/aws/aws-ssm-secret-provider.ts
import { GetParametersCommand, type SSMClient } from '@aws-sdk/client-ssm';
import {
  ExpiredCredentialsError,
  isExpiredCredentialsError,
} from '../../domain/expired-credentials-error.js';
import type { ISecretProvider } from '../../domain/ports/secret-provider.js';
import {
  isSsoSessionExpiredError,
  SsoSessionExpiredError,
} from '../../domain/sso-session-expired-error.js';

const SSM_BATCH_SIZE = 10;

/**
 * {@link ISecretProvider} backed by AWS SSM Parameter Store.
 *
 * Parameters are retrieved with decryption enabled so that
 * SecureString values are returned in plain text.
 *
 * SSM supports fetching up to 10 parameters per request,
 * so names are chunked into batches automatically.
 */
export class AwsSsmSecretProvider implements ISecretProvider {
  private readonly ssmClient: SSMClient;
  private readonly profile?: string;

  constructor(ssmClient: SSMClient, profile?: string) {
    if (!ssmClient) {
      throw new Error('ssmClient cannot be null');
    }
    this.ssmClient = ssmClient;
    this.profile = profile;
  }

  async getSecrets(names: string[]): Promise<Map<string, string>> {
    const result = new Map<string, string>();
    if (names.length === 0) {
      return result;
    }

    for (const name of names) {
      if (!name?.trim()) {
        throw new Error('Secret name cannot be null or whitespace');
      }
    }

    for (let i = 0; i < names.length; i += SSM_BATCH_SIZE) {
      const batch = names.slice(i, i + SSM_BATCH_SIZE);
      try {
        const response = await this.ssmClient.send(
          new GetParametersCommand({
            Names: batch,
            WithDecryption: true,
          }),
        );

        for (const param of response.Parameters ?? []) {
          if (param.Name && param.Value != null) {
            result.set(param.Name, param.Value);
          }
        }
      } catch (error) {
        if (isSsoSessionExpiredError(error)) {
          throw new SsoSessionExpiredError(this.profile, error);
        }
        if (isExpiredCredentialsError(error)) {
          throw new ExpiredCredentialsError(error);
        }
        throw error;
      }
    }

    return result;
  }
}
